shared/src/commonMain/kotlin/io/customer/shared/serializer/CustomAttributeContextualSerializer.kt

```diff
@@ -0,0 +1,108 @@
+package io.customer.shared.serializer
+
+import io.customer.shared.serializer.CustomAttributeContextualSerializer.Companion.DEFAULT_FALLBACK_VALUE
+import io.customer.shared.util.Logger
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.*
+
+/**
+ * Contextual serializer for Kotlin serialization to add support for generic objects. The
+ * serializer assures the parsing is safe and fallbacks to [DEFAULT_FALLBACK_VALUE] where it
+ * fails to process.
+ *
+ * To support [Any] object in the class, the class file should add the following annotation
+ *
+ * [@file:UseContextualSerialization(Any::class)]
+ */
+internal class CustomAttributeContextualSerializer(
+    private val logger: Logger,
+    internal val serializer: CustomAttributeSerializer?,
+) : KSerializer<Any> {
+    private val delegateSerializer = JsonElement.serializer()
+    override val descriptor: SerialDescriptor = delegateSerializer.descriptor
+
+    override fun serialize(encoder: Encoder, value: Any) {
+        val jsonElement = toJsonElementOrNull(value = value).also { json ->
+            if (json == null) {
+                logger.error("Unable to serialize $value, replacing with $DEFAULT_FALLBACK_VALUE")
+            }
+        } ?: JsonPrimitive(value = DEFAULT_FALLBACK_VALUE)
+        encoder.encodeSerializableValue(delegateSerializer, jsonElement)
+    }
+
+    override fun deserialize(decoder: Decoder): Any {
+        val jsonElement = decoder.decodeSerializableValue(delegateSerializer)
+        return toAnyValueOrNull(jsonElement = jsonElement) ?: DEFAULT_FALLBACK_VALUE
+    }
+
+    companion object {
+        const val DEFAULT_FALLBACK_VALUE = "NULL"
+    }
+}
+
+/**
+ * Maps objects to json elements. Basic primitive types and collections are supported. All
+ * unknown object types mapped using [CustomAttributeSerializer] if available.
+ *
+ * @return mapped [JsonElement] if successful; null otherwise.
+ */
+private fun CustomAttributeContextualSerializer.toJsonElementOrNull(value: Any?): JsonElement? {
+    return when (value) {
```

Comment:
What happens for the type `List` will it be considered as `JsonArray`. 

I think we probably need to match the behavior of the serializer with API, for example, from what i can remember, could be wrong we used to make comma separated string for List items in attributes.

Reply:
Yes, it does. I did try to cover maximum possible data types, but I don't exactly remember if I tested this or not 😅 I'll test this again and confirm soon. This should, at least, support all the types we were supporting previously for `CustomAttributes`, and if it doesn't, we'll fix this to work the same way.

Reply:
List was supported, but Arrays were missed 🤐  I [pushed fixes](https://github.com/customerio/mobile-shared/pull/4/files/0ff09da7105435a15877790d4231789175401a0d..fea528e733e60b34068481218f9ca87b42b3b835) for a few more issues I identified while testing. I took some inspiration from [TestCustomAttributesDataSet in our Android repository](https://github.com/customerio/customerio-android/blob/develop/common-test/src/main/java/io/customer/commontest/TestCustomAttributesDataSet.kt) and here are the results:

`CustomAttributes` I tested with:
```
mapOf(
  "string" to "foobar",
  "long" to 394949L,
  "nested" to mapOf("nested-key" to "nested-value"),
  "date" to 1646410611L.unixTimeToDate(),
  "colors-array" to arrayOf("Red", "Green", "Blue"),
  "colors-list" to listOf("Cyan", "Magenta", "Yellow"),
  "dynamic-list" to listOf(1, "Black", User(
      email = "dummy@email.com",
      name = "Dummy",
      isGuest = false,
  )),
  "big-decimal" to java.math.BigDecimal(Double.MAX_VALUE + 1),
  "ENUM-key" to mapOf(
      "Hydration-key" to HabitType.Hydration,
      "TakingBreaks-key" to HabitType.TakingBreaks,
      "FocusTime-key" to HabitType.FocusTime,
  ).toSortedMap(),
  "data" to CustomAttributesTestDataClass(String.random),
)
```

Without custom serializer:
```
{
  "string": "foobar",
  "long": 394949,
  "nested": {
    "nested-key": "nested-value"
  },
  "date": 1646410611000,
  "colors-array": [
    "Red",
    "Green",
    "Blue"
  ],
  "colors-list": [
    "Cyan",
    "Magenta",
    "Yellow"
  ],
  "dynamic-list": [
    1,
    "Black",
    null
  ],
  "big-decimal": 1.7976931348623157E308,
  "ENUM-key": {
    "FocusTime-key": "FocusTime",
    "Hydration-key": "Hydration",
    "TakingBreaks-key": "TakingBreaks"
  },
  "data": null
}
```

With custom serializer:
```
{
  "string": "foobar",
  "long": 394949,
  "nested": {
    "nested-key": "nested-value"
  },
  "date": 1646410611000,
  "colors-array": [
    "Red",
    "Green",
    "Blue"
  ],
  "colors-list": [
    "Cyan",
    "Magenta",
    "Yellow"
  ],
  "dynamic-list": [
    1,
    "Black",
    {
      "email": "dummy@email.com",
      "name": "Dummy",
      "isGuest": false
    }
  ],
  "big-decimal": 1.7976931348623157E308,
  "ENUM-key": {
    "FocusTime-key": "FocusTime",
    "Hydration-key": "Hydration",
    "TakingBreaks-key": "TakingBreaks"
  },
  "data": "lfvppobacw"
}
```

And serializer function looks like
```
override fun serialize(value: Any): Any? {
    return when (value) {
        is User -> {
            mapOf(
                "email" to value.email,
                "name" to value.name,
                "isGuest" to value.isGuest,
            )
        }
        is CustomAttributesTestDataClass -> {
            value.foo
        }
        else -> null
    }
}

```

The only difference I could figure out is, unknown types were apparently parsed to `{}` previously, but are now parsed to `null`. Assuming we now provide custom serializer options, I think this is a good change. Would love to see what you folks think about this.